File: ImplementMinStack/ImplementMinStack.cpp
// ImplementMinStack.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class MinStack
{
public:
    MinStack(int capacity)
        : m_Capacity(capacity)
        , m_pArray(new int[capacity])
        , m_Size(0)
        , m_MinElement(0)
    {}

    ~MinStack()
    {
        if (m_pArray)
            delete[] m_pArray;
    }

    void push(int x)
    {
        if (m_Size == m_Capacity)
            return;

        if (m_Size == 0)
        {
            m_MinElement = x;
            m_pArray[m_Size++] = x;
        }
        else if (x < m_MinElement)
        {
            int temp = x * 2 - m_MinElement;
            m_pArray[m_Size++] = temp;
            m_MinElement = x;
        }
        else
            m_pArray[m_Size++] = x;
    }

    void pop()
    {
        if (m_Size == 0)
            return;

        if (m_pArray[--m_Size] < m_MinElement)
        {
            m_MinElement = m_MinElement * 2 - m_pArray[m_Size];
        }
    }

    int top()
    {
        if (m_pArray[m_Size - 1] < m_MinElement)
        {
            return m_MinElement;
        }
        return m_pArray[m_Size - 1];
    }

    int getMin()
    {
        return m_MinElement;
    }

private:
    int m_Capacity;
    int* m_pArray;
    int m_Size;
    int m_MinElement;
};

int main()
{
    MinStack s(10);
    s.push(3);
    s.push(8);
    s.push(1);
    s.push(2);
    s.push(-1);
    s.push(7);
    s.push(-9);

    std::cout << s.getMin() << std::endl;
    s.pop();
    std::cout << s.top() << std::endl;
    std::cout << s.getMin() << std::endl;
    s.pop();
    s.pop();
    std::cout << s.top() << std::endl;
    std::cout << s.getMin() << std::endl;
    s.pop();
    std::cout << s.top() << std::endl;
    std::cout << s.getMin() << std::endl;
    s.pop();
    std::cout << s.top() << std::endl;
    std::cout << s.getMin() << std::endl;
}

File: Go/ImplementMinStack.go
package main

import "fmt"

type MinStack struct {
	m_Capacity int
	m_Size     int
	m_Array    []int
	m_Min      int
}

func NewStack(capacity int) *MinStack {
	return &MinStack{capacity, 0, make([]int, capacity), 0}
}

func (s *MinStack) Push(x int) {
	if s.m_Size == s.m_Capacity {
		return
	}

	if s.m_Size == 0 {
		s.m_Min = x
		s.m_Array[0] = x
	} else if x < s.m_Min {
		s.m_Array[s.m_Size] = x*2 - s.m_Min
		s.m_Min = x
	} else {
		s.m_Array[s.m_Size] = x
	}

	s.m_Size++
}

func (s *MinStack) Pop() {
	if s.m_Size == 0 {
		return
	}

	n := s.m_Array[s.m_Size-1]
	if n < s.m_Min {
		if s.m_Size != 1 {
			s.m_Min = s.m_Min*2 - n
		}
	}

	s.m_Size--
}

func (s *MinStack) Top() int {
	if s.m_Size == 0 {
		return -1
	}

	n := s.m_Array[s.m_Size-1]
	if n < s.m_Min {
		return s.m_Min
	}

	return n
}

func (s *MinStack) GetMin() int {
	return s.m_Min
}

func main() {
	s := NewStack(10)
	s.Push(3)
	s.Push(8)
	s.Push(1)
	s.Push(2)
	s.Push(-1)
	s.Push(7)
	s.Push(-9)
	fmt.Println(s.GetMin())
	s.Pop()
	fmt.Println(s.Top())
	fmt.Println(s.GetMin())
	s.Pop()
	s.Pop()
	fmt.Println(s.Top())
	fmt.Println(s.GetMin())
	s.Pop()
	fmt.Println(s.Top())
	fmt.Println(s.GetMin())
	s.Pop()
	fmt.Println(s.Top())
	fmt.Println(s.GetMin())
}
